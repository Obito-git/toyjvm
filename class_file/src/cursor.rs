use crate::ParseError;

pub(crate) struct Cursor<'a> {
    it: std::iter::Copied<std::slice::Iter<'a, u8>>,
}

impl<'a> Cursor<'a> {
    pub fn new(data: &'a [u8]) -> Self {
        Self {
            it: data.iter().copied(),
        }
    }

    fn take<const N: usize>(it: &mut impl Iterator<Item = u8>) -> Option<[u8; N]> {
        let mut buf = [0u8; N];
        for b in &mut buf {
            *b = it.next()?;
        }
        Some(buf)
    }

    pub fn u32(&mut self) -> Result<u32, ParseError> {
        Cursor::take::<4>(&mut self.it)
            .map(u32::from_be_bytes)
            .ok_or(ParseError::UnexpectedEof)
    }

    pub fn u16(&mut self) -> Result<u16, ParseError> {
        Cursor::take::<2>(&mut self.it)
            .map(u16::from_be_bytes)
            .ok_or(ParseError::UnexpectedEof)
    }

    pub fn u8(&mut self) -> Result<u8, ParseError> {
        self.it.next().ok_or(ParseError::UnexpectedEof)
    }

    pub fn bytes(&mut self, n: usize) -> Result<Vec<u8>, ParseError> {
        let mut v = Vec::with_capacity(n);
        for _ in 0..n {
            v.push(self.it.next().ok_or(ParseError::UnexpectedEof)?);
        }
        Ok(v)
    }

    pub fn read_exact(&mut self, buf: &mut [u8]) -> Result<(), ParseError> {
        for byte in buf {
            *byte = self.it.next().ok_or(ParseError::UnexpectedEof)?;
        }
        Ok(())
    }
}
