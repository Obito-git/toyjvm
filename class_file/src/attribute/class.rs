use crate::constant_pool::ConstantPool;
use crate::ClassFileErr;
use common::cursor::ByteCursor;
use std::fmt;
use std::fmt::{Display, Formatter};

pub const ATTR_SOURCE_FILE: &[u8] = b"SourceFile";

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ClassAttribute {
    SourceFile { sourcefile_index: u16 },
    Unknown { name_index: u16, info: Vec<u8> },
}

impl<'a> ClassAttribute {
    pub(crate) fn read(
        pool: &ConstantPool,
        cursor: &mut ByteCursor<'a>,
    ) -> Result<Self, ClassFileErr> {
        let attribute_name_index = cursor.u16()?;
        let attribute_length = cursor.u32()? as usize;

        let utf8 = pool.get_utf8(attribute_name_index)?.as_bytes();
        match utf8 {
            ATTR_SOURCE_FILE => Ok(ClassAttribute::SourceFile {
                sourcefile_index: cursor.u16()?,
            }),
            _ => {
                let mut buf = vec![0u8; attribute_length];
                cursor.read_exact(&mut buf)?;
                Ok(ClassAttribute::Unknown {
                    name_index: attribute_name_index,
                    info: buf,
                })
            }
        }
    }
}

impl Display for ClassAttribute {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            ClassAttribute::SourceFile { sourcefile_index } => {
                write!(f, "SourceFile(sourcefile_index: {})", sourcefile_index)
            }
            ClassAttribute::Unknown { name_index, info } => write!(
                f,
                "Unsupported(name_index: {}, data: {} bytes)",
                name_index,
                info.len()
            ),
        }
    }
}
