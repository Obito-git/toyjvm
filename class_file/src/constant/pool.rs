use crate::constant::{ConstantInfo, ConstantTag, DynamicInfo, NameAndTypeInfo, ReferenceInfo};
use crate::error::ClassFileErr;
#[cfg(test)]
use serde::Serialize;

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html#jvms-4.4
#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, PartialEq)]
pub struct ConstantPool {
    pub inner: Vec<ConstantInfo>,
}

impl ConstantPool {
    pub fn get_utf8(&self, idx: &u16) -> Result<&str, ClassFileErr> {
        self.inner
            .get(*idx as usize)
            .ok_or(ClassFileErr::ConstantNotFound(*idx))
            .and_then(|entry| match entry {
                ConstantInfo::Utf8(value) => Ok(value.as_str()),
                e => Err(ClassFileErr::TypeError(
                    *idx,
                    ConstantTag::Utf8,
                    e.get_tag(),
                )),
            })
    }
}

#[cfg(feature = "pretty_print")]
/// Getters that are useful only for pretty printing
impl ConstantPool {
    pub fn get_raw(&self, idx: &u16) -> Result<&ConstantInfo, ClassFileErr> {
        self.inner
            .get(*idx as usize)
            .ok_or(ClassFileErr::ConstantNotFound(*idx))
    }

    pub fn get_integer(&self, idx: &u16) -> Result<&i32, ClassFileErr> {
        self.inner
            .get(*idx as usize)
            .ok_or(ClassFileErr::ConstantNotFound(*idx))
            .and_then(|entry| match entry {
                ConstantInfo::Integer(value) => Ok(value),
                e => Err(ClassFileErr::TypeError(
                    *idx,
                    ConstantTag::Integer,
                    e.get_tag(),
                )),
            })
    }

    pub fn get_class(&self, idx: &u16) -> Result<u16, ClassFileErr> {
        self.inner
            .get(*idx as usize)
            .ok_or(ClassFileErr::ConstantNotFound(*idx))
            .and_then(|entry| match entry {
                ConstantInfo::Class(name_index) => Ok(*name_index),
                e => Err(ClassFileErr::TypeError(
                    *idx,
                    ConstantTag::Class,
                    e.get_tag(),
                )),
            })
    }

    pub fn get_class_name(&self, idx: &u16) -> Result<&str, ClassFileErr> {
        let name_index = self.get_class(idx)?;
        self.get_utf8(&name_index)
    }

    pub fn get_pretty_class_name(&self, idx: &u16) -> Result<String, ClassFileErr> {
        let name_index = self.get_class(idx)?;
        self.get_utf8(&name_index).map(|raw| raw.replace('/', "."))
    }

    pub fn get_methodref(&self, idx: &u16) -> Result<&ReferenceInfo, ClassFileErr> {
        self.inner
            .get(*idx as usize)
            .ok_or(ClassFileErr::ConstantNotFound(*idx))
            .and_then(|entry| match entry {
                ConstantInfo::MethodRef(ref_info) => Ok(ref_info),
                e => Err(ClassFileErr::TypeError(
                    *idx,
                    ConstantTag::MethodRef,
                    e.get_tag(),
                )),
            })
    }

    pub fn get_name_and_type(&self, idx: &u16) -> Result<&NameAndTypeInfo, ClassFileErr> {
        self.inner
            .get(*idx as usize)
            .ok_or(ClassFileErr::ConstantNotFound(*idx))
            .and_then(|entry| match entry {
                ConstantInfo::NameAndType(ref_info) => Ok(ref_info),
                e => Err(ClassFileErr::TypeError(
                    *idx,
                    ConstantTag::NameAndType,
                    e.get_tag(),
                )),
            })
    }

    pub fn get_nat_name(&self, nat: &NameAndTypeInfo) -> Result<&str, ClassFileErr> {
        self.get_utf8(&nat.name_index)
    }

    pub fn get_nat_descriptor(&self, nat: &NameAndTypeInfo) -> Result<&str, ClassFileErr> {
        self.get_utf8(&nat.descriptor_index)
    }

    pub fn get_dyn_info_name(&self, dyn_info: &DynamicInfo) -> Result<&str, ClassFileErr> {
        let nat = self.get_name_and_type(&dyn_info.name_and_type_index)?;
        self.get_nat_name(nat)
    }

    pub fn get_dyn_info_descriptor(&self, dyn_info: &DynamicInfo) -> Result<&str, ClassFileErr> {
        let nat = self.get_name_and_type(&dyn_info.name_and_type_index)?;
        self.get_nat_descriptor(nat)
    }

    pub fn get_method_class_name(&self, method_ref: &ReferenceInfo) -> Result<&str, ClassFileErr> {
        self.get_class_name(&method_ref.class_index)
    }

    pub fn get_method_name(&self, method_ref: &ReferenceInfo) -> Result<&str, ClassFileErr> {
        let nat_index = method_ref.name_and_type_index;
        let nat = self.get_name_and_type(&nat_index)?;
        self.get_nat_name(nat)
    }

    pub fn get_descriptor(&self, ref_info: &ReferenceInfo) -> Result<&str, ClassFileErr> {
        let nat_index = ref_info.name_and_type_index;
        let desc_index = self.get_name_and_type(&nat_index)?;
        self.get_nat_descriptor(desc_index)
    }
}
