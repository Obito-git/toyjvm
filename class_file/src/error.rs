use crate::attribute::AttributeType;
use crate::constant::ConstantTag;
use common::utils::cursor::CursorError;
#[cfg(test)]
use serde::Serialize;
use thiserror::Error;

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum ClassFileErr {
    #[error(transparent)]
    Cursor(#[from] CursorError),
    #[error("")]
    WrongMagic,
    #[error("Expected end of file but found trailing bytes.")]
    TrailingBytes,
    #[error("")]
    UnknownTag(u8),
    #[error("Expected type `{1}` with index `{0}` but found `{2}`")]
    /// First u16 is index, second is expected type, third is actual type
    TypeError(u16, ConstantTag, ConstantTag),
    #[error("Constant with index `{0}` isn't found in constant constant.")]
    ConstantNotFound(u16),
    #[error("Unknown stack frame type {0}.")]
    UnknownStackFrameType(u8),
    #[error("Unknown attribute `{0}.")]
    UnknownAttribute(String),
    #[error("Can't build shared attribute, the `{0}` attribute isn't shared.")]
    AttributeIsNotShared(AttributeType),
    #[error("Invalid method handle kind {0}.")]
    InvalidMethodHandleKind(u8),
}
