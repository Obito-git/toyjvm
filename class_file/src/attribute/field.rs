use crate::attribute::{AttributeType, SharedAttribute};
use crate::constant::pool::ConstantPool;
use crate::error::ClassFormatErr;
use common::utils::cursor::ByteCursor;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FieldAttribute {
    Shared(SharedAttribute),
    ConstantValue(u16),
}

impl<'a> FieldAttribute {
    pub(crate) fn read(
        pool: &ConstantPool,
        cursor: &mut ByteCursor<'a>,
    ) -> Result<Self, ClassFormatErr> {
        let attribute_name_index = cursor.u16()?;
        let _attribute_length = cursor.u32()? as usize;

        let attribute_type = AttributeType::try_from(pool.get_utf8(&attribute_name_index)?)?;
        match attribute_type {
            AttributeType::ConstantValue => Ok(FieldAttribute::ConstantValue(cursor.u16()?)),
            AttributeType::RuntimeVisibleAnnotations
            | AttributeType::Synthetic
            | AttributeType::Deprecated
            | AttributeType::Signature => Ok(FieldAttribute::Shared(SharedAttribute::read(
                attribute_type,
                cursor,
            )?)),
            _ => unimplemented!(),
        }
    }

    #[cfg(feature = "pretty_print")]
    pub(crate) fn fmt_pretty(
        &self,
        ind: &mut common::utils::indent_write::Indented<'_>,
        cp: &ConstantPool,
    ) -> std::fmt::Result {
        use common::pretty_try;
        use std::fmt::Write as _;

        match self {
            FieldAttribute::Shared(shared) => shared.fmt_pretty(ind, cp)?,
            FieldAttribute::ConstantValue(val) => {
                let constant = pretty_try!(ind, cp.get_raw(val));
                writeln!(
                    ind,
                    "ConstantValue: {}",
                    pretty_try!(ind, constant.get_pretty_value(cp, &0))
                )?;
            }
        }

        Ok(())
    }
}
