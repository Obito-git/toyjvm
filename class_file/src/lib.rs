use crate::attribute::class::ClassAttribute;
use crate::constant::pool::ConstantPool;
use crate::error::ClassFileErr;
use common::utils::cursor::ByteCursor;
use constant::ConstantInfo;
use field::FieldInfo;
use method::MethodInfo;
#[cfg(test)]
use serde::Serialize;

pub mod attribute;
pub mod constant;
pub mod error;
pub mod field;
pub mod method;
#[cfg(feature = "pretty_print")]
pub mod print;
// TODO: review all access levels in the crate (methods, fields, modules, structs, etc.)
// TODO: align enums that end with "Info"/"Ref" and "Type"/"Kind" suffixes

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html
/// A rust representation of a Java .class file. All structures in the crates have public only public
/// fields for easier access, because anyway it will be remapped to runtime structures.
///
/// All print related code is behind the `pretty_print` feature flag.
#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub struct ClassFile {
    pub minor_version: u16,
    pub major_version: u16,
    pub cp: ConstantPool,
    pub access_flags: u16,
    pub this_class: u16,
    pub super_class: u16,
    pub interfaces: Vec<u16>,
    pub fields: Vec<FieldInfo>,
    pub methods: Vec<MethodInfo>,
    pub attributes: Vec<ClassAttribute>,
}

impl ClassFile {
    const MAGIC: u32 = 0xCAFEBABE;
    fn validate_magic(val: u32) -> Result<(), ClassFileErr> {
        (val == ClassFile::MAGIC)
            .then_some(())
            .ok_or(ClassFileErr::WrongMagic)
    }
}

impl TryFrom<Vec<u8>> for ClassFile {
    type Error = ClassFileErr;

    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
        let mut cursor = ByteCursor::new(&value);
        let magic = cursor.u32()?;
        ClassFile::validate_magic(magic)?;
        let minor_version = cursor.u16()?;
        let major_version = cursor.u16()?;
        let constant_pool_count = cursor.u16()?;
        let mut constant_pool = Vec::with_capacity((constant_pool_count + 1) as usize);
        constant_pool.push(ConstantInfo::Unused);
        let mut i = 1;
        while i < constant_pool_count {
            let constant = ConstantInfo::read(&mut cursor)?;
            constant_pool.push(constant.clone());
            match constant {
                // described in JVM spec that Long and Double take two entries in the constant pool
                // https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html#jvms-4.4.5
                ConstantInfo::Long(_) | ConstantInfo::Double(_) => {
                    constant_pool.push(ConstantInfo::Unused);
                    i += 2;
                }
                _ => {
                    i += 1;
                }
            }
        }
        let constant_pool = ConstantPool {
            inner: constant_pool,
        };

        let access_flags = cursor.u16()?;
        let this_class = cursor.u16()?;
        let super_class = cursor.u16()?;
        let interfaces_count = cursor.u16()?;
        let mut interfaces = Vec::with_capacity(interfaces_count as usize);
        for _ in 0..interfaces_count {
            interfaces.push(cursor.u16()?);
        }
        let fields_count = cursor.u16()?;
        let mut fields = Vec::with_capacity(fields_count as usize);
        for _ in 0..fields_count {
            fields.push(FieldInfo::read(&constant_pool, &mut cursor)?);
        }
        let methods_count = cursor.u16()?;
        let mut methods = Vec::with_capacity(methods_count as usize);
        for _ in 0..methods_count {
            methods.push(MethodInfo::read(&constant_pool, &mut cursor)?);
        }
        let attributes_count = cursor.u16()?;
        let mut attributes = Vec::with_capacity(attributes_count as usize);
        for _ in 0..attributes_count {
            attributes.push(ClassAttribute::read(&constant_pool, &mut cursor)?);
        }

        if cursor.u8().is_ok() {
            Err(ClassFileErr::TrailingBytes)
        } else {
            Ok(Self {
                minor_version,
                major_version,
                cp: constant_pool,
                access_flags,
                this_class,
                super_class,
                interfaces,
                fields,
                methods,
                attributes,
            })
        }
    }
}

#[cfg(feature = "pretty_print")]
impl std::fmt::Display for ClassFile {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use crate::print::{get_class_javap_like_list, get_class_pretty_java_like_prefix};
        use common::utils::indent_write::Indented;
        use common::{pretty_class_name_try, pretty_try};
        use std::fmt::Write as _;

        const CONSTANT_KIND_WIDTH: usize = 18;
        let mut ind = Indented::new(f);
        writeln!(
            ind,
            "{} {}",
            get_class_pretty_java_like_prefix(self.access_flags),
            pretty_class_name_try!(ind, self.cp.get_class_name(&self.this_class))
        )?;
        ind.with_indent(|ind| {
            writeln!(ind, "minor version: {}", self.minor_version)?;
            writeln!(ind, "major version: {}", self.major_version)?;
            writeln!(
                ind,
                "flags: (0x{:04X}) {}",
                self.access_flags,
                get_class_javap_like_list(self.access_flags)
            )?;
            writeln!(
                ind,
                "this_class: {:<24}//{}",
                format!("#{}", self.this_class),
                pretty_try!(ind, self.cp.get_class_name(&self.this_class))
            )?;
            writeln!(ind, "super_class: #{}", self.super_class)?;
            writeln!(
                ind,
                "interfaces: {}, fields: {}, methods: {}, attributes: {}",
                self.interfaces.len(),
                self.fields.len(),
                self.methods.len(),
                self.attributes.len()
            )?;
            Ok(())
        })?;
        writeln!(ind, "Constant pool:")?;
        ind.with_indent(|ind| {
            let counter_width = self
                .cp
                .inner
                .len()
                .checked_ilog10()
                .map_or(0, |d| d as usize)
                + 2;
            for (i, c) in self.cp.inner.iter().enumerate() {
                if matches!(c, ConstantInfo::Unused) {
                    continue;
                }
                let tag = format_args!("{:<kw$}", c.get_tag(), kw = CONSTANT_KIND_WIDTH);
                write!(ind, "{:>w$} = {} ", format!("#{i}"), tag, w = counter_width)?;
                c.fmt_pretty(ind, &self.cp)?;
            }
            Ok(())
        })?;
        writeln!(ind, "{{")?;
        ind.with_indent(|ind| {
            for (i, field) in self.fields.iter().enumerate() {
                field.fmt_pretty(ind, &self.cp)?;
                if i + 1 < self.methods.len() {
                    writeln!(ind)?;
                }
            }
            Ok(())
        })?;
        writeln!(ind, "}}")?;
        writeln!(ind, "{{")?;
        ind.with_indent(|ind| {
            for (i, method) in self.methods.iter().enumerate() {
                method.fmt_pretty(ind, &self.cp, &self.this_class)?;
                if i + 1 < self.methods.len() {
                    writeln!(ind)?;
                }
            }
            Ok(())
        })?;
        writeln!(ind, "}}")?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::with_settings;
    use rstest::*;
    use std::fs;
    use std::io::BufRead;
    use std::path::{Path, PathBuf};

    const CLASS_SNAPSHOT_PATH: &str = "../snapshots/class_file";
    const DISPLAY_SNAPSHOT_PATH: &str = "../snapshots/display";

    fn to_snapshot_name(path: &Path) -> String {
        let mut iter = path.iter().map(|s| s.to_string_lossy().to_string());
        for seg in iter.by_ref() {
            if seg == "test-classes" {
                break;
            }
        }
        let tail: Vec<String> = iter.collect();
        tail.join("-")
    }

    #[rstest]
    #[trace]
    // Requires `testdata/compile-fixtures.py` to be executed to generate the .class files
    fn parse_class_file(
        #[base_dir = "../target/test-classes"]
        #[files("**/*.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));

        // When
        let class_file = ClassFile::try_from(bytes).unwrap();

        // Then
        /*
        with_settings!(
            {
                snapshot_path => CLASS_SNAPSHOT_PATH,
                prepend_module_to_snapshot => false,
            },
            {
                insta::assert_yaml_snapshot!(to_snapshot_name(&path), class_file);
            }
        );
        */
    }

    #[rstest]
    #[trace]
    // Requires `testdata/compile-fixtures.sh` to be run to generate the .class files
    fn test_display(
        #[base_dir = "../target/test-classes"]
        #[files("**/*.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));

        // When
        let display = format!("{}", ClassFile::try_from(bytes).unwrap());

        // Then
        /*
        with_settings!(
            {
                snapshot_path => DISPLAY_SNAPSHOT_PATH,
                prepend_module_to_snapshot => false,
            },
            {
                insta::assert_snapshot!(to_snapshot_name(&path), display);
            }
        );
         */
    }

    #[rstest]
    // Requires `testdata/compile-fixtures.sh` to be run to generate the .class files
    fn compare_with_javap(
        #[base_dir = "../target/test-classes"]
        #[files("**/*.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));
        let my_display = format!("{}", ClassFile::try_from(bytes).unwrap())
            .as_bytes()
            .lines()
            .collect::<Result<Vec<_>, _>>()
            .unwrap();
        let javap = std::process::Command::new("javap")
            .arg("-v")
            .arg("-p")
            .arg(&path)
            .output()
            .unwrap_or_else(|e| panic!("Can't run javap on file {:?}:{}\n", path, e))
            .stdout
            .lines()
            .collect::<Result<Vec<_>, _>>()
            .unwrap();
        let javap_display = javap[4..javap.len() - 1].to_vec();

        // When && Then
        for (i, (my, javap)) in my_display.iter().zip(javap_display.iter()).enumerate() {
            let my_line_normalized: String = my.chars().filter(|c| !c.is_whitespace()).collect();
            let javap_line_normalized: String =
                javap.chars().filter(|c| !c.is_whitespace()).collect();
            assert_eq!(
                my_line_normalized,
                javap_line_normalized,
                "Mismatch at line {} of file {:?}",
                i + 1,
                path,
            );
        }
    }
}
