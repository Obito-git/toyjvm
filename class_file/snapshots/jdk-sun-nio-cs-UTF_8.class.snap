---
source: class_file/src/lib.rs
expression: display
---
public final class sun.nio.cs.UTF_8 extends sun.nio.cs.Unicode
  minor version: 0
  major version: 68
  flags: (0x0031) ACC_PUBLIC, ACC_FINAL, ACC_SUPER
  this_class: #39                      //sun/nio/cs/UTF_8
  super_class: #10                     //sun/nio/cs/Unicode
  interfaces: 0, fields: 2, methods: 7, attributes: 3
Constant pool:
   #1 = String             #2               // UTF-8
   #2 = Utf8               UTF-8
   #3 = Methodref          #4.#5            // sun/nio/cs/StandardCharsets.aliases_UTF_8:()[Ljava/lang/String;
   #4 = Class              #6               // sun/nio/cs/StandardCharsets
   #5 = NameAndType        #7:#8            // aliases_UTF_8:()[Ljava/lang/String;
   #6 = Utf8               sun/nio/cs/StandardCharsets
   #7 = Utf8               aliases_UTF_8
   #8 = Utf8               ()[Ljava/lang/String;
   #9 = Methodref          #10.#11          // sun/nio/cs/Unicode."<init>":(Ljava/lang/String;[Ljava/lang/String;)V
  #10 = Class              #12              // sun/nio/cs/Unicode
  #11 = NameAndType        #13:#14          // "<init>":(Ljava/lang/String;[Ljava/lang/String;)V
  #12 = Utf8               sun/nio/cs/Unicode
  #13 = Utf8               <init>
  #14 = Utf8               (Ljava/lang/String;[Ljava/lang/String;)V
  #15 = String             #16              // UTF8
  #16 = Utf8               UTF8
  #17 = Class              #18              // sun/nio/cs/UTF_8$Decoder
  #18 = Utf8               sun/nio/cs/UTF_8$Decoder
  #19 = Methodref          #17.#20          // sun/nio/cs/UTF_8$Decoder."<init>":(Ljava/nio/charset/Charset;)V
  #20 = NameAndType        #13:#21          // "<init>":(Ljava/nio/charset/Charset;)V
  #21 = Utf8               (Ljava/nio/charset/Charset;)V
  #22 = Class              #23              // sun/nio/cs/UTF_8$Encoder
  #23 = Utf8               sun/nio/cs/UTF_8$Encoder
  #24 = Methodref          #22.#20          // sun/nio/cs/UTF_8$Encoder."<init>":(Ljava/nio/charset/Charset;)V
  #25 = Methodref          #26.#27          // java/nio/Buffer.arrayOffset:()I
  #26 = Class              #28              // java/nio/Buffer
  #27 = NameAndType        #29:#30          // arrayOffset:()I
  #28 = Utf8               java/nio/Buffer
  #29 = Utf8               arrayOffset
  #30 = Utf8               ()I
  #31 = Methodref          #26.#32          // java/nio/Buffer.position:(I)Ljava/nio/Buffer;
  #32 = NameAndType        #33:#34          // position:(I)Ljava/nio/Buffer;
  #33 = Utf8               position
  #34 = Utf8               (I)Ljava/nio/Buffer;
  #35 = Methodref          #10.#36          // sun/nio/cs/Unicode.contains:(Ljava/nio/charset/Charset;)Z
  #36 = NameAndType        #37:#38          // contains:(Ljava/nio/charset/Charset;)Z
  #37 = Utf8               contains
  #38 = Utf8               (Ljava/nio/charset/Charset;)Z
  #39 = Class              #40              // sun/nio/cs/UTF_8
  #40 = Utf8               sun/nio/cs/UTF_8
  #41 = Methodref          #39.#42          // sun/nio/cs/UTF_8."<init>":()V
  #42 = NameAndType        #13:#43          // "<init>":()V
  #43 = Utf8               ()V
  #44 = Fieldref           #39.#45          // sun/nio/cs/UTF_8.INSTANCE:Lsun/nio/cs/UTF_8;
  #45 = NameAndType        #46:#47          // INSTANCE:Lsun/nio/cs/UTF_8;
  #46 = Utf8               INSTANCE
  #47 = Utf8               Lsun/nio/cs/UTF_8;
  #48 = Methodref          #49.#50          // jdk/internal/access/SharedSecrets.getJavaLangAccess:()Ljdk/internal/access/JavaLangAccess;
  #49 = Class              #51              // jdk/internal/access/SharedSecrets
  #50 = NameAndType        #52:#53          // getJavaLangAccess:()Ljdk/internal/access/JavaLangAccess;
  #51 = Utf8               jdk/internal/access/SharedSecrets
  #52 = Utf8               getJavaLangAccess
  #53 = Utf8               ()Ljdk/internal/access/JavaLangAccess;
  #54 = Fieldref           #39.#55          // sun/nio/cs/UTF_8.JLA:Ljdk/internal/access/JavaLangAccess;
  #55 = NameAndType        #56:#57          // JLA:Ljdk/internal/access/JavaLangAccess;
  #56 = Utf8               JLA
  #57 = Utf8               Ljdk/internal/access/JavaLangAccess;
  #58 = Utf8               Code
  #59 = Utf8               LineNumberTable
  #60 = Utf8               LocalVariableTable
  #61 = Utf8               this
  #62 = Utf8               historicalName
  #63 = Utf8               ()Ljava/lang/String;
  #64 = Utf8               newDecoder
  #65 = Utf8               ()Ljava/nio/charset/CharsetDecoder;
  #66 = Utf8               newEncoder
  #67 = Utf8               ()Ljava/nio/charset/CharsetEncoder;
  #68 = Utf8               updatePositions
  #69 = Utf8               (Ljava/nio/Buffer;ILjava/nio/Buffer;I)V
  #70 = Utf8               src
  #71 = Utf8               Ljava/nio/Buffer;
  #72 = Utf8               sp
  #73 = Utf8               I
  #74 = Utf8               dst
  #75 = Utf8               dp
  #76 = Utf8               MethodParameters
  #77 = Utf8               <clinit>
  #78 = Utf8               SourceFile
  #79 = Utf8               UTF_8.java
  #80 = Utf8               NestMembers
  #81 = Utf8               InnerClasses
  #82 = Utf8               Decoder
  #83 = Utf8               Encoder
{
  public static final sun.nio.cs.UTF_8 INSTANCE;
    descriptor: Lsun/nio/cs/UTF_8;
    flags: (0x0019) ACC_PUBLIC, ACC_STATIC, ACC_FINAL
  
  private static final jdk.internal.access.JavaLangAccess JLA;
    descriptor: Ljdk/internal/access/JavaLangAccess;
    flags: (0x001a) ACC_PRIVATE, ACC_STATIC, ACC_FINAL
  
  public sun.nio.cs.UTF_8();
    descriptor: ()V
    flags: (0x0001) ACC_PUBLIC
    Code: 
      stack=3, locals=1, args_size=1
         0: aload_0                 
         1: ldc           #1                  // String UTF-8
         3: invokestatic  #3                  // Method sun/nio/cs/StandardCharsets.aliases_UTF_8:()[Ljava/lang/String;
         6: invokespecial #9                  // Method sun/nio/cs/Unicode."<init>":(Ljava/lang/String;[Ljava/lang/String;)V
         9: return                  
      LineNumberTable:
        line 65: 0
        line 66: 9
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0       10     0  this   Lsun/nio/cs/UTF_8;
  
  public java.lang.String historicalName();
    descriptor: ()Ljava/lang/String;
    flags: (0x0001) ACC_PUBLIC
    Code: 
      stack=1, locals=1, args_size=1
         0: ldc           #15                 // String UTF8
         2: areturn                 
      LineNumberTable:
        line 69: 0
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0        3     0  this   Lsun/nio/cs/UTF_8;
  
  public java.nio.charset.CharsetDecoder newDecoder();
    descriptor: ()Ljava/nio/charset/CharsetDecoder;
    flags: (0x0001) ACC_PUBLIC
    Code: 
      stack=3, locals=1, args_size=1
         0: new           #17                 // class sun/nio/cs/UTF_8$Decoder
         3: dup                     
         4: aload_0                 
         5: invokespecial #19                 // Method sun/nio/cs/UTF_8$Decoder."<init>":(Ljava/nio/charset/Charset;)V
         8: areturn                 
      LineNumberTable:
        line 73: 0
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0        9     0  this   Lsun/nio/cs/UTF_8;
  
  public java.nio.charset.CharsetEncoder newEncoder();
    descriptor: ()Ljava/nio/charset/CharsetEncoder;
    flags: (0x0001) ACC_PUBLIC
    Code: 
      stack=3, locals=1, args_size=1
         0: new           #22                 // class sun/nio/cs/UTF_8$Encoder
         3: dup                     
         4: aload_0                 
         5: invokespecial #24                 // Method sun/nio/cs/UTF_8$Encoder."<init>":(Ljava/nio/charset/Charset;)V
         8: areturn                 
      LineNumberTable:
        line 77: 0
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0        9     0  this   Lsun/nio/cs/UTF_8;
  
  static final void updatePositions(java.nio.Buffer, int, java.nio.Buffer, int);
    descriptor: (Ljava/nio/Buffer;ILjava/nio/Buffer;I)V
    flags: (0x0018) ACC_STATIC, ACC_FINAL
    Code: 
      stack=3, locals=4, args_size=4
         0: aload_0                 
         1: iload_1                 
         2: aload_0                 
         3: invokevirtual #25                 // Method java/nio/Buffer.arrayOffset:()I
         6: isub                    
         7: invokevirtual #31                 // Method java/nio/Buffer.position:(I)Ljava/nio/Buffer;
        10: pop                     
        11: aload_2                 
        12: iload_3                 
        13: aload_2                 
        14: invokevirtual #25                 // Method java/nio/Buffer.arrayOffset:()I
        17: isub                    
        18: invokevirtual #31                 // Method java/nio/Buffer.position:(I)Ljava/nio/Buffer;
        21: pop                     
        22: return                  
      LineNumberTable:
        line 82: 0
        line 83: 11
        line 84: 22
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0       23     0  src    Ljava/nio/Buffer;
           0       23     1  sp     I
           0       23     2  dst    Ljava/nio/Buffer;
           0       23     3  dp     I
  
  public boolean contains(java.nio.charset.Charset);
    descriptor: (Ljava/nio/charset/Charset;)Z
    flags: (0x1041) ACC_PUBLIC, ACC_BRIDGE, ACC_SYNTHETIC
    Code: 
      stack=2, locals=2, args_size=2
         0: aload_0                 
         1: aload_1                 
         2: invokespecial #35                 // Method sun/nio/cs/Unicode.contains:(Ljava/nio/charset/Charset;)Z
         5: ireturn                 
      LineNumberTable:
        line 60: 0
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0        6     0  this   Lsun/nio/cs/UTF_8;
    MethodParameters:
      Name                             Flags
      <no name>                        synthetic 
  
  static {};
    descriptor: ()V
    flags: (0x0008) ACC_STATIC
    Code: 
      stack=2, locals=0, args_size=0
         0: new           #39                 // class sun/nio/cs/UTF_8
         3: dup                     
         4: invokespecial #41                 // Method "<init>":()V
         7: putstatic     #44                 // Field INSTANCE:Lsun/nio/cs/UTF_8;
        10: invokestatic  #48                 // Method jdk/internal/access/SharedSecrets.getJavaLangAccess:()Ljdk/internal/access/JavaLangAccess;
        13: putstatic     #54                 // Field JLA:Ljdk/internal/access/JavaLangAccess;
        16: return                  
      LineNumberTable:
        line 62: 0
        line 86: 10
}
SourceFile: "UTF_8.java"
NestMembers:
  sun/nio/cs/UTF_8$Encoder
  sun/nio/cs/UTF_8$Decoder
InnerClasses:
  private static #82= #17 of #39;             // Decoder=class sun/nio/cs/UTF_8$Decoder of class sun/nio/cs/UTF_8
  private static final #83= #22 of #39;       // Encoder=class sun/nio/cs/UTF_8$Encoder of class sun/nio/cs/UTF_8
